packages/application-shell-connectors/src/components/application-context/application-context.tsx

```diff
@@ -191,27 +200,52 @@ function withApplicationContext<
   };
 }
 
-// Forward-compatibility with React Hooks 🎉
-const useApplicationContext = React.useContext
-  ? <
-      SelectedContext extends {} = {},
-      AdditionalEnvironmentProperties extends {} = {}
-    >(
-      selector?: (
-        context: TApplicationContext<AdditionalEnvironmentProperties> | {}
-      ) => SelectedContext
-    ):
-      | SelectedContext
-      | TApplicationContext<AdditionalEnvironmentProperties>
-      | {} => {
-      const applicationContext = React.useContext(Context);
-      return selector ? selector(applicationContext) : applicationContext;
-    }
-  : () => {
-      throw new Error(
-        `React hooks do not seem to be available yet in the installed React version "${React.version}". Please check the React Hooks documentation for more info: https://reactjs.org/hooks.`
-      );
-    };
+// Use function overloading to declare two possible signatures with two
+// distict return types, based on the selector function argument.
+function useApplicationContextHook<
+  AdditionalEnvironmentProperties extends {} = {}
+>(): TApplicationContext<AdditionalEnvironmentProperties>;
+function useApplicationContextHook<
+  SelectedContext,
+  AdditionalEnvironmentProperties extends {} = {}
+>(
+  selector: (
+    context: TApplicationContext<AdditionalEnvironmentProperties>
+  ) => SelectedContext
+): SelectedContext;
+
+// Then implement the function. Typescript will pick the appropriate signature
+// based on the function arguments.
+function useApplicationContextHook<
+  SelectedContext,
+  AdditionalEnvironmentProperties extends {} = {}
+>(
+  selector?: (
+    context: TApplicationContext<AdditionalEnvironmentProperties>
+  ) => SelectedContext
+) {
+  // Forward-compatibility with React Hooks 🎉
+  if (!React.useContext) {
+    throw new Error(
+      `React hooks do not seem to be available yet in the installed React version "${React.version}". Please check the React Hooks documentation for more info: https://reactjs.org/hooks.`
+    );
+  }
+
+  const context = React.useContext(Context);
+  // Because of the way the ApplicationShell configures the Context.Provider,
+  // we ensure that, when we read from the context, we always get actual
+  // context object and not the initial value.
+  // Therefore, we can safely cast the value to be out `TApplicationContext` type.
+  const applicationContext = context as TApplicationContext<
+    AdditionalEnvironmentProperties
+  >;
+  return selector ? selector(applicationContext) : applicationContext;
+}
+
+// This is a workaround to trick babel/rollup to correctly export the function.
+// Most likely the problem arises with the use of overloading.
+// See related issue: https://github.com/babel/babel/issues/8361
+const useApplicationContext = useApplicationContextHook;
```

Comment:
This fixes this problem: https://github.com/commercetools/merchant-center-application-kit/pull/740#issuecomment-501858918